### PR TITLE
autoround_support_qbits_backend

### PR DESCRIPTION
**Pull request by zhewang1-intc**
as title. 
add a demo file in auto_round dir for better review experience.
note: autoGPTQ may use qbits as default CPU inference kernel in the future, pls refer to this pr https://github.com/AutoGPTQ/AutoGPTQ/pull/660 and issue https://github.com/AutoGPTQ/AutoGPTQ/issues/655,
I think this pr could be a temp-workaround, once pr660 get merged, we can import quantlinear from autoGPTQ directly.

### CODE REVIEW

**Comment by zhewang1-intc** [comment]
@wenhuach21 @WeiweiZhang1 could you pls take a look?

File: auto_round/qlinear_qbits.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            "scales",
+            torch.zeros(
+                (math.ceil(infeatures / self.group_size), outfeatures),
+                dtype=weight_dtype,
+            ),
+        )
+        self.register_buffer(
+            "g_idx",
+            torch.tensor(
+                [i // self.group_size for i in range(infeatures)], dtype=torch.int32),
```
Comment (wenhuach21) [inline auto_round/qlinear_qbits.py]: remove g_idx as never used

File: auto_round/qlinear_qbits.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        )
+        if bias:
+            self.register_buffer("bias", torch.zeros(
+                (outfeatures), dtype=torch.float))
+        else:
+            self.bias = None
+
+        self.kernel_switch_threshold = kernel_switch_threshold
+
+        self.trainable = trainable
```
Comment (wenhuach21) [inline auto_round/qlinear_qbits.py]: better create a qbits dir in auto_round_extenstion

File: auto_round/export/export_to_autoround/export.py [hunk 3]
```diff
@@ -89,15 +90,18 @@ def dynamic_QuantLienar_for_packing(backend, bits, group_size):
...
+            from auto_round_extension.cuda.qliner_triton import QuantLinear
+            return QuantLinear
 
     else:
         assert False, f"only support gptq and autoround backend"
 
 
 @register_format("auto_round")
-def save_quantized_as_autoround(output_dir, inplace=True, backend="autoround:exllamav2", **kwargs):
+def save_quantized_as_autoround(output_dir, inplace=True, backend="autoround:qbits", **kwargs):
```
Comment (wenhuach21) [inline auto_round/export/export_to_autoround/export.py]: could qbits supports loading exllmav2 format model, I prefer to use exllamav2 as the default as it's commonly used in the community

File: auto_round/auto_quantizer.py [hunk 4]
```diff
@@ -311,7 +312,9 @@ def convert_model(self, model: nn.Module):
         return model
 
     def _dynamic_import_inference_linear(self, bits, backend):
-        if bits == 4 and self.exllama2_available and "exllamav2" in backend:
+        if "qbits" in backend:
+            return qlinear_qbits.QuantLinear
```
Comment (wenhuach21) [inline auto_round/auto_quantizer.py]: add itrex in the requirements in someplace to support qbits

File: qbits_backend_demo.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+path = "/home/wangzhe/.cache/huggingface/hub/models--facebook--opt-125m/snapshots/27dcfa74d334bc871f3234de431e71c6eeba5dd6"
+model = AutoModelForCausalLM.from_pretrained(
+    path, torch_dtype="auto", trust_remote_code=True)
+tokenizer = AutoTokenizer.from_pretrained(path, trust_remote_code=True)
+
+
+bits, group_size, sym = 4, 128, False
+# device:Optional["auto", None, "hpu", "cpu", "cuda"]
+autoround = AutoRound(model, tokenizer, bits=bits, group_size=group_size,
+                      sym=sym, nsamples=32, seqlen=32, device="cpu", iters=1, scale_dtype="fp32")
```
Comment (wenhuach21) [inline qbits_backend_demo.py]: the default scale_dtype is fp16, is it the must for qbits? please remove this file before the pr merging

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿*.exe
+*.dll
```
Comment (WeiweiZhang1) [inline .gitignore]: please remove this file

File: auto_round/qlinear_qbits.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+# limitations under the License.
+
+import math
+
+import numpy as np
+import torch
+import torch.nn as nn
+import transformers
+from intel_extension_for_transformers import qbits  # with QBits kernels ()
+from logging import getLogger
```
Comment (WeiweiZhang1) [inline auto_round/qlinear_qbits.py]: from autoround.utils import logger

File: auto_round/qlinear_qbits.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+logger = getLogger(__name__)
+
+BITS_DTYPE_MAPPING = {
+    4: "int4_clip",
+    8: "int8",
+}
+
+
+def convert_dtype_torch2str(dtype):
```
Comment (WeiweiZhang1) [inline auto_round/qlinear_qbits.py]: from autoround.utils import convert_dtype_torch2str

File: auto_round/qlinear_qbits.py [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+        self.trainable = trainable
+
+    def post_init(self):
+        assert self.qweight.device.type == "cpu"
+        if self.bias is not None:
+            self.bias = self.bias.to(dtype=torch.float32)
+
+        default_idx = torch.tensor(
+            [i // self.group_size for i in range(self.infeatures)], dtype=torch.int32)
+        is_desc_act = not torch.all(torch.eq(default_idx, self.g_idx))
```
Comment (WeiweiZhang1) [inline auto_round/qlinear_qbits.py]: the handling of desc_act is also not required by autoround

**Review by WeiweiZhang1 — APPROVED** [review]

File: auto_round/export/export_to_autoround/export.py (hunk 3 above)
Comment (zhewang1-intc) [reply auto_round/export/export_to_autoround/export.py]: fix in [9287524](https://github.com/intel/auto-round/pull/145/commits/92875249b5a0dfb8602be5ebf2ee07fbced8dce2)

File: auto_round/qlinear_qbits.py (hunk 2 above)
Comment (zhewang1-intc) [reply auto_round/qlinear_qbits.py]: fix in [5a68b52](https://github.com/intel/auto-round/pull/145/commits/5a68b5252cd711cb2fca3ed9297662449b5fbd4c)

File: auto_round/qlinear_qbits.py (hunk 1 above)
Comment (zhewang1-intc) [reply auto_round/qlinear_qbits.py]: fix in bacad15

File: auto_round/auto_quantizer.py (hunk 4 above)
Comment (zhewang1-intc) [reply auto_round/auto_quantizer.py]: fix in [ca2aa6e](https://github.com/intel/auto-round/pull/145/commits/ca2aa6efc4cbffb4ea874b8447aca47857c0e935)

File: .gitignore (hunk 6 above)
Comment (zhewang1-intc) [reply .gitignore]: fix in bacad15,
btw i keep some necessary ignore files/dirs (e.g. auto_round.egg-info/)

File: auto_round/qlinear_qbits.py (hunk 7 above)
Comment (zhewang1-intc) [reply auto_round/qlinear_qbits.py]: fix in bacad15

File: auto_round/qlinear_qbits.py (hunk 8 above)
Comment (zhewang1-intc) [reply auto_round/qlinear_qbits.py]: fix in bacad15

File: auto_round/qlinear_qbits.py (hunk 9 above)
Comment (zhewang1-intc) [reply auto_round/qlinear_qbits.py]: fix in bacad15

File: qbits_backend_demo.py (hunk 5 above)
Comment (zhewang1-intc) [reply qbits_backend_demo.py]: yes, already changed to default value.

File: auto_round/auto_quantizer.py [hunk 10]
```diff
@@ -316,8 +317,11 @@ def convert_model(self, model: nn.Module):
         return model
 
     def _dynamic_import_inference_linear(self, bits, backend):
-        if bits == 4 and self.exllama2_available and "exllamav2" in backend:
-            from auto_round_extension.cuda.qliner_exllamav2 import QuantLinear
+        if bits == 4 and "exllamav2" in backend:
+            if self.exllama2_available:
+                from auto_round_extension.cuda.qliner_exllamav2 import QuantLinear
+            else:
+                return qlinear_qbits.QuantLinear
```
Comment (wenhuach21) [inline auto_round/auto_quantizer.py]: as qbits is only for cpu, could we check the model device to determine which backend is used?

File: auto_round/export/export_to_autoround/export.py [hunk 11]
```diff
@@ -89,8 +90,11 @@ def dynamic_QuantLienar_for_packing(backend, bits, group_size):
         return QuantLinear
     ##export all use trition, inference use exllamav2
     elif "autoround" in backend or "auto-round" in backend or "auto_round" in backend:
-        from auto_round_extension.cuda.qliner_triton import QuantLinear
-        return QuantLinear
+        if device.type == "cpu":
+            return qlinear_qbits.QuantLinear
```
Comment (wenhuach21) [inline auto_round/export/export_to_autoround/export.py]: During the packing phase, the model is always on the CPU. If the packing of qbits differs, it is better to use the standard format, as qbits are compatible with the standard. However, the reverse is not true.

File: auto_round_extension/qbits/qlinear_qbits.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+                    qweight[row] |= intweight[j] << (3 * (j - i) + 2)
+                i += 10
+                row += 1
+            else:
+                raise NotImplementedError("Only 2,3,4,8 bits are supported.")
+
+        qweight = qweight.astype(np.int32)
+        self.qweight = torch.from_numpy(qweight)
+
+        zeros -= 1
```
Comment (wenhuach21) [inline auto_round_extension/qbits/qlinear_qbits.py]: remove -1 and +1 as it's  a bug as reported in autogptq

File: auto_round_extension/qbits/qlinear_qbits.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+                (outfeatures), dtype=torch.float))
+        else:
+            self.bias = None
+
+        self.kernel_switch_threshold = kernel_switch_threshold
+
+        self.trainable = trainable
+
+    def post_init(self):
+        assert self.qweight.device.type == "cpu"
```
Comment (wenhuach21) [inline auto_round_extension/qbits/qlinear_qbits.py:95]: We still need to support models packed with ExLlama2, so I guess this function should be called in  auto_quantier.py

File: auto_round/auto_quantizer.py (hunk 10 above)
Comment (zhewang1-intc) [reply auto_round/auto_quantizer.py]: in [9cbc870](https://github.com/intel/auto-round/pull/145/commits/9cbc870c4f380419101dab923630383ee5ca4eb9) we will import default exllamav2 qlinear and then check whether the qweight tensor on cpu device, if true we will switch the qlinear to qbis_linear.

File: auto_round/export/export_to_autoround/export.py (hunk 11 above)
Comment (zhewang1-intc) [reply auto_round/export/export_to_autoround/export.py]: yes, your are right, i use the default triton linear when exporting model and this dosen't affect model output.

File: auto_round_extension/qbits/qlinear_qbits.py (hunk 12 above)
Comment (zhewang1-intc) [reply auto_round_extension/qbits/qlinear_qbits.py]: turn to default triton linear packing which doesn't have zp +/- 1 behavior.

File: auto_round_extension/qbits/qlinear_qbits.py (hunk 13 above)
Comment (zhewang1-intc) [reply auto_round_extension/qbits/qlinear_qbits.py:95]: done in 9cbc870

**Comment by zhewang1-intc** [comment]
acc looks good.
model facebook/opt-125m
quant param:
```
autoround = AutoRound(model, tokenizer, bits=bits, group_size=group_size,
                      sym=sym, nsamples=256, seqlen=512, device="cpu", iters=200)
```
asym ACC
|    Tasks     |Version|Filter|n-shot|  Metric  | Value |   |Stderr|
|--------------|------:|------|-----:|----------|------:|---|-----:|
|lambada_openai|      1|none  |     0|perplexity|29.0069|±  |1.0685|
|              |       |none  |     0|acc       | 0.3621|±  |0.0067|

sym ACC
|    Tasks     |Version|Filter|n-shot|  Metric  | Value |   |Stderr|
|--------------|------:|------|-----:|----------|------:|---|-----:|
|lambada_openai|      1|none  |     0|perplexity|28.6612|±  |1.0599|
|              |       |none  |     0|acc       | 0.3621|±  |0.0067|

raw model ACC
|    Tasks     |Version|Filter|n-shot|  Metric  | Value |   |Stderr|
|--------------|------:|------|-----:|----------|------:|---|-----:|
|lambada_openai|      1|none  |     0|perplexity|26.0200|±  |0.9382|
|              |       |none  |     0|acc       | 0.3786|±  |0.0068|

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1,11 @@
+﻿.vs
+.vscode
+__pycache__
+auto_round.egg-info/
+build/*
```
Comment (wenhuach21) [inline .gitignore:5]: don't forget to delete this file

File: qbits_backend_demo.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+# device:Optional["auto", None, "hpu", "cpu", "cuda"]
+autoround = AutoRound(model, tokenizer, bits=bits, group_size=group_size,
+                      sym=sym, nsamples=32, seqlen=32, device="cpu", iters=1)
+autoround.quantize()
+output_dir = "./tmp_autoround"
+autoround.save_quantized(output_dir, format="auto_round")
+
+
+from auto_round.auto_quantizer import *
+from auto_round_extension.qbits.qlinear_qbits import *
```
Comment (wenhuach21) [inline qbits_backend_demo.py]: you may add an unit test instead of demo file

Comment (zhewang1-intc) [reply qbits_backend_demo.py]: remove it as we have qbits ut in itrex.

**Comment by zhewang1-intc** [comment]
int2 autoround quantize acc:
asym 
|    Tasks     |Version|Filter|n-shot|  Metric  | Value  |   |Stderr|
|--------------|------:|------|-----:|----------|-------:|---|-----:|
|lambada_openai|      1|none  |     0|perplexity|184.9410|±  |8.5206|
|              |       |none  |     0|acc       |  0.1758|±  |0.0053|

sym
|    Tasks     |Version|Filter|n-shot|  Metric  | Value  |   |Stderr|
|--------------|------:|------|-----:|----------|-------:|---|-----:|
|lambada_openai|      1|none  |     0|perplexity|184.9410|±  |8.5206|
|              |       |none  |     0|acc       |  0.1758|±  |0.0053|

**Comment by zhewang1-intc** [comment]
Intel/Qwen2-7B-int4-inc QBits backend acc
|    Tasks     |Version|Filter|n-shot|  Metric  |Value |   |Stderr|
|--------------|------:|------|-----:|----------|-----:|---|-----:|
|lambada_openai|      1|none  |     0|perplexity|3.5641|±  |0.0739|
|              |       |none  |     0|acc       |0.7246|±  |0.0062|
